Add visibility:[all|public|private] search filter

mrnugget opened the pull request:
This is supposed to fix #8344. I've tested it on my local machine and it works as it should.

But I don't know if the code is in the right place and where to add which tests. cc @rvantonder and @sourcegraph/web

tsenart commented on cmd/frontend/graphqlbackend/visibility.go:
```diff
@@ -0,0 +1,20 @@
+type repoVisibility string
+
+const (
+	All     repoVisibility = "all"
```
`Any` visibility sounds better to me than `All` visibility.

mrnugget replied:
Great suggestion!

tsenart commented on cmd/frontend/db/repos.go:
```diff
@@ -441,6 +447,12 @@ func (*repos) listSQL(opt ReposListOptions) (conds []*sqlf.Query, err error) {
+		conds = append(conds, sqlf.Sprintf("NOT private"))
+	}
+	if opt.OnlyPrivate {
+		conds = append(conds, sqlf.Sprintf("private"))
```
Do we want / need an index?

unknwon replied:
We don't have index for `fork` and `arhicved`, so maybe not right now.

mrnugget replied:
Adding them.

unknwon commented on cmd/frontend/db/repos.go:
```diff
@@ -253,6 +253,12 @@ type ReposListOptions struct {
+	// NoPrivate excludes private repositories from the list.
+	NoPrivate bool
+
+	// OnlyPrivate excludes public repositories from the list.
```
nit:
```suggestion
	// OnlyPrivate excludes non-private repositories from the list.
```

For the sake of GitLab's "internal" visibility.

tsenart commented:
> We don't have index for fork and arhicved, so maybe not right now.

That doesn't mean we shouldn't.

lguychard commented on web/src/search/input/interactive/filters.ts:
```diff
@@ -80,4 +80,5 @@ export const FilterTypeToProseNames: Record<FilterType, string> = {
     content: 'Content',
     patterntype: 'Pattern type',
     index: 'Indexed repos',
+    visibility: 'Repository visiblity',
```
I think you need to add this filter to `FiniteFilterType`, `finiteFilters` for it to display correctly in interactive mode (cc @attfarhan to confirm).

mrnugget replied:
Added it in 7c1a17d. Let me know if that's right/wrong @attfarhan

attfarhan replied:
Looks right @mrnugget

rvantonder commented on cmd/frontend/graphqlbackend/visibility.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+func parseVisibility(s string) repoVisibility {
+	switch s {
```
You may want to throw in a `strings.ToLower(s)` here. We do this for some other fields (somewhere, I'm not sure off the top of my head).

rvantonder commented on cmd/frontend/graphqlbackend/visibility.go:
```diff
@@ -0,0 +1,20 @@
+package graphqlbackend
```
Let's put this file in `internal/search/query`. Our directory structure for search isn't perfect, but this would live better there, where I'm trying to concentrate the types. For example, the `patterntype` stuff lives there too.

rvantonder commented on cmd/frontend/graphqlbackend/search.go:
```diff
@@ -525,6 +532,8 @@ func resolveRepositories(ctx context.Context, op resolveRepoOp) (repoRevisions,
 			OnlyForks:    op.onlyForks,
 			NoArchived:   op.noArchived,
 			OnlyArchived: op.onlyArchived,
+			NoPrivate:    op.onlyPublic,
+			OnlyPrivate:  op.onlyPrivate,
```
Yeah, the alias of `NoPrivate` -> `onlyPublic` works for now, under the assumption that we don't support `visibility:internal` (and `internal` will be part of `onlyPublic` I assume). When I get to implementing `select` the need for two booleans here should go away and all of repo selection will become an expression.

Aside: do you know whether we record [`internal` repo labels](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-repository-visibility#about-internal-repositories)? Have we thought about supporting that? Is it possible $customer might want this?

mrnugget replied:
> Aside: do you know whether we record [`internal` repo labels](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-repository-visibility#about-internal-repositories)?

I'm pretty sure we don't record them. They also seem to be a GitHub Enterprise exclusive feature and since we treat GH and GHE kinda the same right now, I'd assume that supporting GHE-only feature would require more work than simply persisting another attribute.

rvantonder commented:
Also: changelog entry :)

mrnugget commented:
I moved the file, added the two suggested fixes, added a CHANGELOG entry, documented the new filter in the search docs, and added an E2E test.

For the e2e test I added a private repository to the `sourcegraph` org, specifically only for tests, added the `sourcegraph/team` and `sourcegraph-bot`.

(After multiple hours of trying to get the setup working) the e2e test works locally. It doesn't seem to run on Buildkite because it cannot seem to find the private repository in the external service setup. I assume that's because the token used in our `buildkite-agents` doesn't belong to `sourcegraph-bot`. So I'm creating a new token and try that.

rvantonder commented on web/src/e2e/e2e.test.ts:
```diff
@@ -362,6 +363,34 @@ describe('e2e test suite', () => {
             // await percySnapshot(page, 'Search results file')
         })
 
+        test('Search visibility:private|public', async () => {
```
Just a general question: why this `e2e` test file rather than `web/src/regression/search.test.ts`? (I'm actually not really what distinction we draw here). 

The ones in `web/src/regression/search.test.ts` have a bunch of search tests I added that I want to migrate to the CLI-based testing we've been floating. The private/public repo test would be a good fit for that too.

mrnugget replied:
The `web/src/regression` test suite doesn't run on every PR/master-build. It only runs when started manually in the release process every month.

If you want, I can move them. But for now I don't think it matters and

> I want to migrate to the CLI-based testing we've been floating

this is probably what we should aim for.

rvantonder replied:
Got it. No it doesn't matter, was just asking.

> this is probably what we should aim for.

Yeah... we neeeed it.
